Enhanced generateDepictionMatching2DStructure functionality

ptosco opened the pull request:
This PR enhances the `generateDepictionMatching2DStructure` functionality.

- `generateDepictionMatching2DStructure` can be used with `referencePattern` smaller than reference to only use part of a scaffold<br/>Previously, `referencePattern` should be strictly the same size of `reference`. While `referencePattern` cannot obviously be larger than `reference`, it can be useful to only freeze part of a reference without having to edit the reference itself
- adds `generateDepictionMatching2DStructure` overload to pass a `matchVect` instead of doing a substructure match<br/>For example, one can already have a `matchVect` from a previous substructure match and doing it again would be a waste of time.<br/>Plus, the scaffold could match the molecule in multiple places and one might have already somehow sorted degeneracy and want to pass explicit indices.
- adds `allowRGroups` parameter to enable using a scaffold bearing R groups as reference.<br/>This is particularly useful to leverage `generateDepictionMatching2DStructure` with scaffold bearing R groups.<br/>Currently, `generateDepictionMatching2DStructure` will fail with molecules bearing implicit hydrogens where R groups are, as they effectively fail the substructure match.<br/>The algorithm in this PR will choose the most specific match available, allowing a minimum of hydrogens as needed for each molecule.<br/>This is especially useful for RDKitJS as it offers very limited possibilities to edit the scaffolds, but it allows saving some boilerplate also from C++/Python/Java.

Note that the Python function `GenerateDepictionMatching2DStructure()` now returns a value : the matchVect corresponding to the substructure match between the atoms in the molecules and the atoms in the reference.

greglandrum commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
 }
 
+namespace detail {
+class MatchScorer {
```
There's no documentation for either what the class does or how it does it.
I'm pretty sure I get it, but if you're going to add something like this, please document it somewhere.

ptosco replied:
Good point, I had completely forgotten about that helper class. I have moved it to `SubstructUtils.cpp` and exposed two functions that use that class in `SubstructUtils.h` as they may useful to others.
I would also like exposing them to Python but as there is no `rdSubstructMatch` module I am not sure where would be the best module to host them (`rdchem`? Suggestions very welcome).

greglandrum commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
+  auto &matchVectRef = matchVect;
   if (referencePattern) {
-    if (reference.getNumAtoms(true) != referencePattern->getNumAtoms(true)) {
+    if (referencePattern->getNumAtoms(true) > reference.getNumAtoms(true)) {
```
`getNumAtoms(true)` is the same as `getNumAtoms()`, might as well go ahead and use the simpler (and more commonly used) form.

ptosco replied:
Yeah, that was just a copy/paste from previous code. Done.

greglandrum commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
+          "<= reference");
     }
     RDKit::MatchVectType refMatchVect;
     RDKit::SubstructMatch(reference, *referencePattern, refMatchVect);
```
Seems like you would want to apply the same `allowRGroups` logic to matching the pattern to the reference too, right? Then you could really easily use a single molecule from a RGD set as the reference and the SMARTS definition of the core as the pattern.

ptosco replied:
Good point, that's certainly useful. Done.

greglandrum commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
+  const RDKit::ROMol &query =
+      (referencePattern ? *referencePattern : reference);
+  if (allowRGroups) {
+    std::unique_ptr<RDKit::ROMol> molHs(RDKit::MolOps::addHs(mol));
```
I would avoid calling `addHs()` (which has a definite runtime cost) unless you know it's necessary. The minimal version of that test is to ensure that the pattern actually contains a dummy atom. 
An alternate, which would make the whole thing faster for cases where you either don't have R groups or they don't need to match to Hs, but would be a bit slower if you both have R groups and they need to match Hs, would be to start by calling `SubstructMatch()` as before and then only following this path if it's necessary (i.e. `SubstructMatch()` didn't return anything and `allowRGroups` is true.

ptosco replied:
Another good point, done.

greglandrum commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
+      (referencePattern ? *referencePattern : reference);
+  if (allowRGroups) {
+    std::unique_ptr<RDKit::ROMol> molHs(RDKit::MolOps::addHs(mol));
+    if (molHs) {
```
At the moment you silently fail if this evaluates to false.
There currently isn't any circumstance under which `addHs()` can return nullptr, so I don't think the test is necessary at all, but if you want to be extra careful, maybe switch to a `CHECK_INVARIANT()`

greglandrum commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
+      detail::MatchScorer matchScorer(*molHs, query);
+      auto matches = SubstructMatch(*molHs, query);
+      if (matches.empty()) {
+        allowRGroups = false;
```
It doesn't seem like this branch, or setting `allowRGroups` to false, is necessary

ptosco replied:
See above comment.

greglandrum commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
+                           const RDKit::MatchVectType &bMatch) {
+              return (matchScorer.score(aMatch) < matchScorer.score(bMatch));
+            });
+        if (matchScorer.score(matchVect) > 1.0) {
```
as I mentioned before, I don't really understand the match scorer, so I don't get why this test is here. I'm pretty sure you need to do something if the test fails thought, right?

ptosco replied:
I have refactored that code and removed that `if` clause.

greglandrum commented on Code/GraphMol/Depictor/Wrap/testDepictor.py:
```diff
@@ -236,6 +237,149 @@ def testConstrainedCoords(self):
+  6  8  1  0
+  7  9  1  0
+  7  5  1  0
+M  END)RES"""
```
I guess the `)RES` here is a copy/paste wart?

ptosco replied:
Oh, yes. Fixed. (Great that it didn't choke the `MolBlock` parser!).

greglandrum commented on Code/GraphMol/Depictor/testDepictor.cpp:
```diff
@@ -1064,6 +1064,208 @@ void testGithub2027() {
+      << "-----------------------\n Test "
+         "generateDepictionMatching2DStructure with allowRGroups"
+      << std::endl;
+  std::string templateMolblock = R"RES(
```
same comment here on constructing the molecule more simply

ptosco replied:
Done.

ptosco commented on Code/GraphMol/Depictor/RDDepictor.cpp:
```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
   }
-  RDDepict::compute2DCoords(mol, &coordMap, false /* canonOrient */,
-                            true /* clearConfs */, 0, 0, 0, false, forceRDKit);
+  if (!allowRGroups) {
```
No, because `allowRGroups` can be set to `false` in the previous `if` clause. That may happen in the unlikely (but not impossible case) that a nasty user has specified a SMARTS query with implicit Hs (lowercase `h`) that would fail against the molecules with explicit Hs, but would succeed against the original molecule with implicit hs.

ptosco commented on Code/GraphMol/Depictor/testDepictor.cpp:
```diff
@@ -1064,6 +1064,208 @@ void testGithub2027() {
+      << "-----------------------\n Test "
+         "generateDepictionMatching2DStructure with refPattern and matchVect"
+      << std::endl;
+  std::string indazoleMolblock = R"RES(
+     RDKit          2D
+
+  9 10  0  0  0  0  0  0  0  0999 V2000
+   -6.0878    2.4335    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
+   -7.3867    1.6835    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
+   -7.3867    0.1833    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
+   -6.0878   -0.5666    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
+   -4.7887    0.1833    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
+   -4.7887    1.6835    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
+   -3.4897   -0.5664    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
+   -2.1906    1.6833    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
+   -2.1906    0.1835    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
+  1  2  2  0
+  2  3  1  0
+  3  4  2  0
+  4  5  1  0
+  5  6  2  0
+  6  1  1  0
+  8  9  2  0
+  6  8  1  0
+  7  9  1  0
+  7  5  1  0
+M  END)RES";
+  std::unique_ptr<ROMol> indazoleRef(MolBlockToMol(indazoleMolblock));
```
Believe it or not, I didn't know of the `_ctab` macro. Noted :)

ptosco commented:
@greglandrum This is now ready for review. The last commit only affects the RDKitJS build.

greglandrum commented on Code/GraphMol/Substruct/SubstructUtils.h:
```diff
@@ -27,6 +27,16 @@ RDKIT_SUBSTRUCTMATCH_EXPORT bool chiralAtomCompat(const Atom* a1,
+//! This postprocesses the passed substruct matches and returns
+//! the match that has the largest number of non-hydrogen atoms
+//! in correspondence of terminal dummy atoms
+RDKIT_SUBSTRUCTMATCH_EXPORT std::vector<MatchVectType>::const_iterator
```
returning an iterator here feels somehow wrong: it complicates client code and raises the question: "what does it mean if I increment/decrement the result of this function?"

How about simplifying things by returning a `const MatchVectType &`?

ptosco replied:
Yeah, of course, that's a leftover of when the function was not publicly accessible, that needs to be changed now. Done.

greglandrum commented on Code/GraphMol/Depictor/RDDepictor.h:
```diff
@@ -149,17 +156,56 @@ RDKIT_DEPICTOR_EXPORT unsigned int compute2DCoordsMimicDistMat(
 
 */
-RDKIT_DEPICTOR_EXPORT void generateDepictionMatching2DStructure(
+RDKIT_DEPICTOR_EXPORT RDKit::MatchVectType generateDepictionMatching2DStructure(
```
Please document the return type of this function.

ptosco replied:
It seems documented to me; what am I missing?

greglandrum replied:
where is it documented? I don't see it in this comment:
https://github.com/ptosco/rdkit/blob/allowRGroups/Code/GraphMol/Depictor/RDDepictor.h#L140

ptosco replied:
Oh, crap, I put in the wrong comment :) I'll fix that now.
https://github.com/ptosco/rdkit/blob/6f09b2a808f41ee9d82394110b519a1f6a28d708/Code/GraphMol/Depictor/RDDepictor.h#L202

greglandrum commented on Code/GraphMol/Depictor/Wrap/rdDepictor.cpp:
```diff
@@ -101,18 +101,35 @@ unsigned int Compute2DCoordsMimicDistmat(
+  auto matchVect = RDDepict::generateDepictionMatching2DStructure(
+      mol, reference, confId, referencePattern, acceptFailure, forceRDKit,
+      allowRGroups);
+  PyObject *atomMap = PyTuple_New(matchVect.size());
```
What you have here isn't wrong, so this is more of a suggestion, but this code would be simpler if you used `python::tuple` instead of the raw Python C API

ptosco replied:
Done.

greglandrum commented on Code/GraphMol/Depictor/Wrap/testDepictor.py:
```diff
@@ -202,6 +202,7 @@ def test5SamplingMimic3D(self):
         # ofile.write(mb)
         # ofile.close()
 
+    #@unittest.skip
```
you can just remove this

ptosco replied:
Yep, forgotten that.

greglandrum reviewed (changes requested):
a few changes requested

ptosco commented:
@greglandrum do you have a suggestion where to expose to Python `sortMatchesByDegreeOfCoreSubstitution` and `getMostSubstitutedCoreMatch`? There isn't a rdSubstructUtils module at the moment, shall I create one and put them there?

greglandrum commented:
> @greglandrum do you have a suggestion where to expose to Python `sortMatchesByDegreeOfCoreSubstitution` and `getMostSubstitutedCoreMatch`? There isn't a `rdSubstructUtils` module at the moment, shall I create one and put them there?

I don't think we need another module for that. How about just putting it in `rdmolops` (the rest of the basic substructure matching ends up coming in with `from rdkit import Chem` so it seems ok if this does too)

ptosco commented:
> I don't think we need another module for that. How about just putting it in `rdmolops` (the rest of the basic substructure matching ends up coming in with `from rdkit import Chem` so it seems ok if this does too)

Yep, I think that works well. I'll do that now.

greglandrum commented:
> > I don't think we need another module for that. How about just putting it in `rdmolops` (the rest of the basic substructure matching ends up coming in with `from rdkit import Chem` so it seems ok if this does too)
> 
> Yep, I think that works well. I'll do that now.

I assume you want to do that as part of this PR or should I go ahead and merge this one and you can add that in a future PR?

ptosco commented:
@greglandrum Yep, I am about to commit, I am almost done with writing unit tests.

ptosco commented:
@greglandrum Got there, eventually. I added Python tests for the wrapper and I realized that `sortMatchesByDegreeOfCoreSubstitution` was not working. The reason is that `ScoreMatchesByDegreeOfCoreSubstitution` cached scores keyed on their pointer: while this works for `std::min_elem`, of course it doesn't for sorting which needs to reshuffle items: in that case the same pointers will point to different values as the vector is reshufled. So I rewrote `ScoreMatchesByDegreeOfCoreSubstitution` without the `std::unordered_map` in a way that works and is also more efficient as it does not involve `unordered_map` lookups.
To play safe then I also added C++ tests.

greglandrum commented:
@ptosco : it looks like the test failures here are real

ptosco commented:
> @ptosco : it looks like the test failures here are real

Indeed. I'll take a look later.

ptosco commented:
@greglandrum This should be good to go once tests pass.
